### PR TITLE
Eric/preferences

### PR DESCRIPTION
**Pull request by estrattonbailey**

### CODE REVIEW

File: src/lib/labeling/const.ts
Comment (estrattonbailey) [inline src/lib/labeling/const.ts:1]: This stuff was only used by preferences, so I co-located it

File: src/state/models/ui/preferences.ts
Comment (estrattonbailey) [inline src/state/models/ui/preferences.ts:1]: Remaining items here will be removed as we update the things that rely on them

File: src/state/models/ui/saved-feeds.ts
Comment (estrattonbailey) [inline src/state/models/ui/saved-feeds.ts:1]: This will be removed after we replace our My Feeds page stuff

File: src/state/queries/feed.ts [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+export const useFeedSourceInfoQueryKey = ({uri}: {uri: string}) => [
+  'getFeedSourceInfo',
+  uri,
+]
+
+const feedSourceNSIDs = {
+  feed: 'app.bsky.feed.generator',
+  list: 'app.bsky.graph.list',
+}
```
Comment (estrattonbailey) [inline src/state/queries/feed.ts:40]: We don't have a clear source for these yet?

File: src/state/queries/preferences/index.ts [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+export const usePreferencesQueryKey = ['getPreferences']
+
+export function usePreferencesQuery() {
+  const {agent} = useSession()
+  return useQuery({
+    queryKey: usePreferencesQueryKey,
+    queryFn: async () => {
+      const res = await agent.getPreferences()
+      const preferences: UsePreferencesQueryResponse = {
```
Comment (estrattonbailey) [inline src/state/queries/preferences/index.ts:31]: I've set defaults for all (or most at least) values here, and updated types to match

File: src/state/queries/preferences/index.ts [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+    void,
+    unknown,
+    {labelGroup: ConfigurableLabelGroup; visibility: LabelPreference}
+  >({
+    mutationFn: async ({labelGroup, visibility}) => {
+      await agent.setContentLabelPref(labelGroup, visibility)
+      // triggers a refetch
+      await queryClient.invalidateQueries({
+        queryKey: usePreferencesQueryKey,
+      })
```
Comment (estrattonbailey) [inline src/state/queries/preferences/index.ts:111]: Rn I'm refetching on every mutation, I figured a shadow cache was overkill for these mutations. We could use RQ cache too.

File: src/state/queries/preferences/moderation.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+  spam: 'hide',
+  impersonation: 'hide',
+}
+
+export const ILLEGAL_LABEL_GROUP: LabelGroupConfig = {
+  id: 'illegal',
+  title: 'Illegal Content',
+  warning: 'Illegal Content',
+  values: ['csam', 'dmca-violation', 'nudity-nonconsensual'],
+}
```
Comment (estrattonbailey) [inline src/state/queries/preferences/moderation.ts:42]: These and the other three are unused?

File: src/state/queries/preferences/moderation.ts [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  impersonation: {
+    id: 'impersonation',
+    title: 'Impersonation',
+    subtitle: 'Accounts falsely claiming to be people or orgs',
+    warning: 'Impersonation',
+    values: ['impersonation'],
+  },
+}
+
+export function getModerationOpts({
```
Comment (estrattonbailey) [inline src/state/queries/preferences/moderation.ts:124]: Replacement for `rootStore.preferences.moderationOpts`, but isn't used anywhere yet

File: src/state/queries/preferences/types.ts [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  feedViewPrefs: BskyPreferences['feedViewPrefs']['home']
+  /**
+   * User thread-view prefs, including newer fields that may not be typed yet.
+   */
+  threadViewPrefs: ThreadViewPreferences
+  userAge: number | undefined
+  feeds: Required<BskyPreferences['feeds']> & {
+    unpinned: string[]
+  }
+}
```
Comment (estrattonbailey) [inline src/state/queries/preferences/types.ts:41]: Cleaned up some types here to include the labs values and computed values done in the hook

File: src/state/queries/preferences/util.ts [hunk 7]
```diff
@@ -0,0 +1,16 @@
+import {LabelPreference} from '@atproto/api'
+
+/**
+ * Content labels previously included 'show', which has been deprecated in
+ * favor of 'ignore'. The API can return legacy data from the database, and
+ * we clean up the data in `usePreferencesQuery`.
+ *
+ * @deprecated
+ */
+export function temp__migrateLabelPref(
```
Comment (estrattonbailey) [inline src/state/queries/preferences/util.ts:10]: Just made this even more obvious

File: src/view/com/auth/create/CreateAccount.tsx [hunk 8]
```diff
@@ -70,13 +78,26 @@ export const CreateAccount = observer(function CreateAccountImpl({
           onboardingDispatch,
           createAccount,
         })
+
+        setBirthDate({birthDate: model.birthDate})
+
+        if (IS_PROD(model.serviceUrl)) {
+          setSavedFeeds(DEFAULT_PROD_FEEDS)
+        }
```
Comment (estrattonbailey) [inline src/view/com/auth/create/CreateAccount.tsx:86]: Side effects, including setting default saved feeds

File: src/view/com/feeds/FeedSourceCard.tsx [hunk 9]
```diff
@@ -16,6 +16,151 @@ import * as Toast from 'view/com/util/Toast'
...
 import {logger} from '#/logger'
 import {useModalControls} from '#/state/modals'
+import {
+  usePreferencesQuery,
+  useSaveFeedMutation,
+  useRemoveFeedMutation,
+} from '#/state/queries/preferences'
+import {useFeedSourceInfoQuery} from '#/state/queries/feed'
+
+export const NewFeedSourceCard = observer(function FeedSourceCardImpl({
```
Comment (estrattonbailey) [inline src/view/com/feeds/FeedSourceCard.tsx:26]: Didn't want to migrate all usages yet, so built this alongside the old one for now

File: src/view/screens/PreferencesHomeFeed.tsx [hunk 10]
```diff
@@ -16,21 +15,31 @@ import {CenteredView} from 'view/com/util/Views'
...
+}: {
+  enabled: boolean
+  initialValue: number
+}) {
   const pal = usePalette('default')
-  const [value, setValue] = useState(
-    store.preferences.homeFeed.hideRepliesByLikeCount,
-  )
+  const [value, setValue] = useState(initialValue)
+  const {mutate: setFeedViewPref} = useSetFeedViewPreferencesMutation()
```
Comment (estrattonbailey) [inline src/view/screens/PreferencesHomeFeed.tsx:32]: This is a little jumpy as the API requests succeeds and state updates, need to come back here later

File: src/view/screens/PreferencesHomeFeed.tsx [hunk 11]
```diff
@@ -176,9 +223,26 @@ export const PreferencesHomeFeed = observer(function PreferencesHomeFeedImpl({
             </Text>
             <ToggleButton
               type="default-light"
-              label={store.preferences.homeFeed.hideQuotePosts ? 'No' : 'Yes'}
-              isSelected={!store.preferences.homeFeed.hideQuotePosts}
-              onPress={store.preferences.toggleHomeFeedHideQuotePosts}
+              label={
+                variables?.hideQuotePosts ??
+                preferences?.feedViewPrefs?.hideQuotePosts
```
Comment (estrattonbailey) [inline src/view/screens/PreferencesHomeFeed.tsx:228]: Local optimistic updates here

File: src/view/screens/SavedFeeds.tsx [hunk 12]
```diff
@@ -43,150 +51,178 @@ const HITSLOP_BOTTOM = {
...
+  isPinned: boolean
 }) {
   const pal = usePalette('default')
-  const isPinned = item.isPinned
+  const queryClient = useQueryClient()
+  const {isPending: isPinPending, mutateAsync: pinFeed} = usePinFeedMutation()
+  const {isPending: isUnpinPending, mutateAsync: unpinFeed} =
+    useUnpinFeedMutation()
+  const {isPending: isMovePending, mutateAsync: setSavedFeeds} =
+    useSetSaveFeedsMutation()
```
Comment (estrattonbailey) [inline src/view/screens/SavedFeeds.tsx:162]: No optimistic updates here bc I don't have a shared context to use `variables` and I didn't want to complicate things with the query cache. We can migrate to that if we need to, but for now this works well. Pending states disable buttons while async req is pending.

File: src/state/queries/preferences/moderation.ts (hunk 4 above)
Comment (pfrazee) [reply src/state/queries/preferences/moderation.ts:42]: Yeah don't worry about that, there's a cleanup pass that needs to happen on the labeling layer to get everything 100%

**Review by pfrazee — APPROVED** [review]
LGTM, merge whenevs

**Pull request merged**